src/libpmemfile-posix/data.c

```diff
@@ -0,0 +1,1332 @@
+	    &offset);
+
+	if (block != NULL)
+		file->block_pointer_cache = block;
+
+	return block;
+}
+
+/*
+ * find_block -- look up block metadata with the highest offset
+ * lower than or equal to the offset argument
+ */
+static struct pmemfile_block *
+find_block(struct pmemfile_vinode *vinode, uint64_t off)
+{
+	return (void *)(uintptr_t)ctree_find_le_unlocked(vinode->blocks, &off);
+}
+
+/*
+ * vinode_destroy_data_state -- destroys file state related to data
+ *
+ * This is used as a callback passed to cb_push_front, that is why the pfp
+ * argument is used.
+ */
+void
+vinode_destroy_data_state(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
+{
+	(void) pfp;
+
+	if (vinode->blocks) {
+		ctree_delete(vinode->blocks);
+		vinode->blocks = NULL;
+	}
+
+	memset(&vinode->first_free_block, 0, sizeof(vinode->first_free_block));
+}
+
+/*
+ * file_allocate_block_data -- allocates new block data.
+ * The block metadata must be already allocated, and passed as the block
+ * pointer argument.
+ */
+static void
+file_allocate_block_data(PMEMfilepool *pfp,
+		struct pmemfile_block *block,
+		size_t count,
+		bool use_usable_size)
+{
+	ASSERT(count > 0);
+	ASSERT(count % FILE_PAGE_SIZE == 0);
+
+	uint32_t size;
+
+	if (pmemfile_posix_block_size != 0) {
+		ASSERT(pmemfile_posix_block_size <= MAX_BLOCK_SIZE);
+		ASSERT(pmemfile_posix_block_size % FILE_PAGE_SIZE == 0);
+
+		size = (uint32_t)pmemfile_posix_block_size;
+	} else {
+		if (count <= MAX_BLOCK_SIZE)
+			size = (uint32_t)count;
+		else
+			size = (uint32_t)MAX_BLOCK_SIZE;
+	}
+
+	block->data = TX_XALLOC(char, size, POBJ_XALLOC_NO_FLUSH);
+	if (use_usable_size) {
+		size_t usable = pmemobj_alloc_usable_size(block->data.oid);
+		ASSERT(usable >= size);
+		if (usable > MAX_BLOCK_SIZE)
+			size = MAX_BLOCK_SIZE;
+		else
+			size = (uint32_t)page_rounddown(usable);
+	}
+
+#ifdef DEBUG
+	/* poison block data */
+	void *data = D_RW(block->data);
+	VALGRIND_ADD_TO_TX(data, size);
+	pmemobj_memset_persist(pfp->pop, data, 0x66, size);
+	VALGRIND_REMOVE_FROM_TX(data, size);
+	VALGRIND_DO_MAKE_MEM_UNDEFINED(data, size);
+#endif
+
+	block->size = size;
+
+	block->flags = 0;
+}
+
+static bool
+is_append(struct pmemfile_vinode *vinode, struct pmemfile_inode *inode,
+		uint64_t offset, uint64_t size)
+{
+	if (inode->size >= offset + size)
+		return false; /* not writing past file size */
+
+	struct pmemfile_block *block = find_last_block(vinode);
+
+	/* Writing past the last allocated block? */
+
+	if (block == NULL)
+		return true;
+
+	return (block->offset + block->size) < (offset + size);
+}
+
+static uint64_t
+overallocate_size(uint64_t count)
+{
+	if (count <= 4096)
+		return 16 * 1024;
+	else if (count <= 64 * 1024)
+		return 256 * 1024;
+	else if (count <= 1024 * 1024)
+		return 4 * 1024 * 1024;
+	else if (count <= 64 * 1024 * 1024)
+		return 64 * 1024 * 1024;
+	else
+		return count;
+}
+
+static void
+vinode_allocate_interval(PMEMfilepool *pfp, struct pmemfile_vinode *vinode,
```

Comment:
Some comments and/or ascii art would be really helpful here for understanding how holes are managed.